### DISCUSSION TITLE
[Feat Req]: Support negative range with singlestat gauge

### DISCUSSION THREAD

**Discussion by StianOvrevage** [grafana/grafana#36970] [Dashboards, Panels & Visualizations]
Please see attached screenshot.
1. Gauge is not showing at all for the negative value, even if it is within the defined range.
2. The colors of the singlestat value are the opposite of the gauge areas. It seems to me that the gauge is correctly colored but the singlestat value itself is backwards.

![grafana singlestat errors](https://cloud.githubusercontent.com/assets/6630005/15690134/6406f072-2782-11e6-8973-e39f7903e1d5.png)

**Comment by GlobularSet** [comment]
I would love to have these feature for the singlestat panel, any idea on where this sits in your queue?

**Comment by hijklmno** [comment]
Would love to have this feature as well. We are gauging our phone signal strength, which is in negative dBm.

**Comment by mohierf** [comment]
The same for me. I am graphing WiFI and Bluetooth RSSI and what I get is:
![image](https://user-images.githubusercontent.com/2932687/31849161-5dd40234-b63e-11e7-99fd-b31c50c26c6b.png)

![image](https://user-images.githubusercontent.com/2932687/31849165-6f9c406c-b63e-11e7-8bdc-ad68f9ab505c.png)

Both gauges have the same configuration but not the same rendering ...

**Comment by dualarrow** [comment]
I have been able to get negative values to load, but the problem seems to be not that its a negative value as such, but that it has decimal places in it.
![image](https://user-images.githubusercontent.com/33943292/33190549-d41c918a-d0e8-11e7-9033-6f48b6c76e21.png)
This was with a MySql datasource and the sql for this is as follows

SELECT
  UNIX_TIMESTAMP(time_ts) as time_sec,
  -floor(solar.PowerReal_P_Sum) as value,
  "Power Usage" as metric
FROM solar
WHERE $__timeFilter(time_ts)
ORDER BY time_ts ASC

What made the guage work for me was the addition of the floor() function.

**Comment by GlobularSet** [comment]
Thanks dualarrow!

I can confirm that this is an issue only for negative values with decimals.  I was able to work around this by force dropping precision in the Options of the panel by setting 'Decimal' to 0

That should work for all data sources (I couldn't floor since this isn't yet supported in our Influx data source).  Cheers!

**Comment by benjamind** [comment]
The issue is in `getDecimalsForValue(value)` specifically this line:

```
    var dec = -Math.floor(Math.log(delta) / Math.LN10);
```

Math.log of negative number is NaN and hence the rest of the code for the auto calculation of decimal count fails. You can workaround by setting the decimal count manually, and non-zero values do work, it just can't auto calculate.

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
